[FIRRTL] Support mixed const operands in firrtl ops

trilorez opened the pull request:
This pr contains the following changes:
- For primitive ops, if all operands to an op are `const`, the result is `const`. If any operand is non-`const`, the result is non-`const`.
- Allow `const` to non-`const` flow for `strictconnect`
- Disallow `const` registers
- Cache purely non-`const` aggregate types for quick type comparisons
- `fold` functions update the op result type to allow for non-'const' to 'const' type propagation
  - This isn't fully implemented for folding to a constant attribute yet since constant ops don't have 'const' result types yet, which will be updated in a later pr
  - Some canonicalization tests have been added but most canonicalization functionality involving `const` will be covered in the later constant op pr
  - This change means `fold` functions might modify the op and return a result, which can be seen as bending the rules about what `fold` functions are allowed to do. I'm open to alternative approaches.

trilorez commented:
After rebasing my upstream branches and consulting with @dtzSiFive, I've updated this pr to also allow const decay in `ref.define` similar to `strictconnect`.

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLDialect.cpp:
```diff
@@ -84,6 +85,55 @@ Operation *FIRRTLDialect::materializeConstant(OpBuilder &builder,
+// those result type changes.
+struct ReinferResultTypes
+    : public mlir::OpInterfaceRewritePattern<mlir::InferTypeOpInterface> {
+  ReinferResultTypes(MLIRContext *context)
```
This seems a little heavy-weight to be run on every op every time it changes?

IIRC we have a cheaper version of this that avoids the need for the vector/etc, which might be cheaper to use for those ops.  Don't know it has an interface though, so maybe nevermind.

trilorez replied:
This was the best option I could find that is compatible with how the canonicalizer works. I'm definitely open to other ideas.

trilorez replied:
The issue is that mutating an operation during a fold is only allowed to affect that operation. But if an operation's result type has changed to const, downstream consumers of that const value might now infer a const result as well.

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLDialect.cpp:
```diff
@@ -84,6 +85,55 @@ Operation *FIRRTLDialect::materializeConstant(OpBuilder &builder,
+    }
+
+    bool anyChanged = false;
+    for (size_t i = 0, e = inferredResultTypes.size(); i != e; ++i) {
```
Or .. maybe `llvm::equal(inferredResultTypes, op->getResults())` or so?

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLFolds.cpp:
```diff
@@ -849,14 +883,14 @@ LogicalResult EQPrimOp::canonicalize(EQPrimOp op, PatternRewriter &rewriter) {
-      return getLhs();
+    if (rhsCst->isZero() && mixedConstTypes(getLhs().getType(), getType()) &&
+        mixedConstTypes(getRhs().getType(), getType()))
+      return foldResult(*this, getLhs());
```
Could we table-gen wrap our fold methods with logic that calls this for us?

Mostly, this means we have to be "good" about always doing this when writing any fold logic, which seems easy to overlook.  Just a thought.

trilorez replied:
I was hoping to find a way to automatically do this but couldn't find anything. If you have any ideas that'd be great.

trilorez replied:
Thanks for the ideas. I think it's worth trying some sort of tablegen op interface approach.

dtzSiFive replied:
!!

Looks like fold /is/ allowed to change the operation in-place, and in that case should return the (new?) operation?

From https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-the-fold-method :

> ///  2. They can mutate the operation in place, without changing anything else
> ///     in the IR. In this case, return the operation itself.

TIL, would this help? Have foldResult return the operation instead of the result Value?

trilorez replied:
I don't think you can actually return the operation, `OpFoldResult` is declared as:
```cpp
class OpFoldResult : public PointerUnion<Attribute, Value>
```

Anyway, so far I've gotten tablegen to optionally define a `fold` method if `hasFolder` is `true`, that calls a `constFold` function meant to be implemented by the user then does all of the result checking. This currently works for single result ops but it would be cool if I could get it to work for single and mult-result ops, then we could put it on `FIRRTLOp` and not have to worry about needing to opt into it.

trilorez replied:
Alright, I've updated `FIRRTLOp` to automatically handle folding type changes.

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -2399,6 +2419,26 @@ LogicalResult RefDefineOp::verify() {
   return success();
 }
 
+static ParseResult parseConstDecay(OpAsmParser &parser, Type &destType,
```
Thanks for this, much appreciate not having to always write/read a lot of extra types! (and less verbose IR is good for all)

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -2399,6 +2419,26 @@ LogicalResult RefDefineOp::verify() {
+                            Type srcType) {
+  p << destType;
+  if (srcType != destType)
+    p << ", " << srcType;
```
Does `operator<<` do the best/most appropriate thing here?  It sure might, but since already crushing it with the optional type printing/parsing, thought I'd ask.  re:qualified/dialect prefix/etc.

trilorez replied:
Hm, in practice it seems to work. I'll look more into this, maybe the `qualified` is irrelevant here

dtzSiFive replied:
Sorry, I was just curious since I couldn't immediately tell.  It probably is fine, but indeed unsure if qualified is important now that you mention it.  Handle as you see fit :+1: .

trilorez replied:
I hunted down the MLIR code that deals with `qualified`. Based on that code, the usage here is meaningless since I'm calling `operator<<`, which calls `printType` which prints the whole type. 

If `qualified` is used in a non-custom context, MLIR generates a printer that calls `printType`. If `qualified`  isn't used, a `printStrippedAttrOrType` call is generated.

dtzSiFive reviewed (commented):
Thanks for all this work, looking good so far.

Gotta run, added a bunch of comments/thoughts/suggestions.

Will try to circle back to have feedback on the big things like the approach to fold/infer/canonicalize.

trilorez commented on include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:
```diff
@@ -85,6 +92,8 @@ def AllOneConstantOp : Constraint<CPred<"$0.getDefiningOp<ConstantOp>() && $0.ge
 def TypeWidthAdjust32 : NativeCodeCall<
   "$_builder.getI32IntegerAttr($0.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel() - $1.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel())">;
 
+def NonConstType : Constraint<CPred<"!isConst($0.getType())">>;
```
Oops, must have been left over after various rebasings

trilorez replied:
Removed

trilorez commented on include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:
```diff
@@ -373,7 +373,7 @@ def RegResetOp : ReferableDeclOp<"regreset", [Forceable]> {
   let arguments = (
     ins ClockType:$clockVal, AnyResetType:$resetSignal,
-        AnyRegisterType:$resetValue,
+        FIRRTLBaseType:$resetValue,
```
`AnyRegisterType`, i.e. the type of a register, cannot be `const`, but the `resetValue` can be `const`. In fact, in the future we will restrict it to _only_ be `const`.

dtzSiFive replied:
Oh, of course.  Great!

trilorez commented on include/circt/Dialect/FIRRTL/FIRRTLTypes.h:
```diff
@@ -177,8 +183,10 @@ class FIRRTLBaseType
 /// compared have any outer flips that encode FIRRTL module directions (input or
 /// output), these should be stripped before using this method.
+/// If `strict` is `true`, `srcFType` must be identical `destFType` except that
+/// 'const' sources can be connected to non-'const' sinks.
```
I initially wrote a separate version, but it seemed like a lot of general logic duplication. But yeah, having a bunch of bools isn't great so maybe that's the better way to go.

trilorez replied:
Updated to a separate function

trilorez commented on include/circt/Dialect/FIRRTL/FIRRTLTypesImpl.td:
```diff
@@ -263,6 +267,10 @@ def BundleImpl : FIRRTLImplType<"Bundle", [FieldIDTypeInterface]> {
     /// Return a 'const' or non-'const' version of this type.
     BundleType getConstType(bool isConst);
 
+    /// Return a non-'const' version of this type with any 'const' types
```
That's probably a possibility. Maybe we should do a general "FIRRTLHardwareType" interface for all of these similar dispatch examples.

trilorez commented on include/circt/Dialect/FIRRTL/FIRRTLOps.h:
```diff
@@ -55,6 +55,9 @@ bool isConstant(Value value);
 /// constant values
 bool isConst(Type type);
 
+// Returns true if the provided types are equal except for constness
```
Updated

trilorez commented:
I realized that we were still allowing nested const fields in the reg ops, so I've added a `hasConst` function. Originally we talked about putting `hasConst` in recursive type properties, which we could still do, but I implemented it by just checking if the type itself is const or if it doesn't equal it's purely non-const type.

darthscsi commented on lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp:
```diff
@@ -734,7 +734,14 @@ void IMConstPropPass::visitOperation(Operation *op) {
   foldResults.reserve(op->getNumResults());
-  if (failed(op->fold(operandConstants, foldResults))) {
+
+  // Since folding can affect the result types, try folding on a temporary clone
```
This is probably pretty expensive.

darthscsi commented on lib/Dialect/FIRRTL/FIRRTLUtils.cpp:
```diff
@@ -44,7 +44,7 @@ void circt::firrtl::emitConnect(ImplicitLocOpBuilder &builder, Value dst,
   }
 
   // If the types are the exact same we can just connect them.
```
comment out of date.

darthscsi commented on include/circt/Dialect/FIRRTL/FIRRTLStatements.td:
```diff
@@ -67,10 +67,10 @@ def StrictConnectOp : FIRRTLOp<"strictconnect", [SameTypeOperands, FConnectLike]
 
   let assemblyFormat =
-    "$dest `,` $src  attr-dict `:` qualified(type($dest))";
+    "$dest `,` $src  attr-dict `:` custom<ConstDecay>(type($dest), type($src))";
```
Strict is pretty strict, it might be better to just cast away the const.

darthscsi commented on include/circt/Dialect/FIRRTL/FIRRTLStatements.td:
```diff
@@ -50,7 +50,7 @@ def ConnectOp : FIRRTLOp<"connect", [FConnectLike]> {
   let hasCanonicalizer = true;
 }
 
-def StrictConnectOp : FIRRTLOp<"strictconnect", [SameTypeOperands, FConnectLike]> {
```
It would be worth writing the constraint here rather than customizing the verifier.  Then it can apply to RefDefineOp also.

darthscsi commented:
Please break this up.  Your list of things done is a good start.  Reg and connect changes seem easily broken out.

Fold needs some attention.  The IMCP hack is likely painful.  Do we really need to do this in the folders?  This seems like a trivial canonicalization, but a dangerous folder.

trilorez commented:
> Reg and connect changes seem easily broken out.

Will do.

> Do we really need to do this in the folders? This seems like a trivial canonicalization, but a dangerous folder.

The main issue is we can't fold a const operand value to a non-const result, since MLIR's canonicalizer rejects mismatched types [here](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Transforms/Utils/FoldUtils.cpp#L276). 

An alternative is not allowing mixed const operands in ops where the result type constness depends on input constness and heavily using const-casts. This gives the benefit of keeping strictconnect strict. 

The drawbacks are verbosity and added complexity to passes. Ops that lookup the defining ops of their operands might need to look through a cast. Wherever constants are created, a cast might need to be inserted.

You can see an older attempt of the cast approach on [this branch](https://github.com/llvm/circt/compare/dev/trilorez/firrtl-const2-constants#diff-2f5ae5c6302ac27ddc66a2d0fc6a98dbd1a96d1a226c0db48909a9bec69244e0).

Which approach seems more palatable?
